src/Libraries/Microsoft.Extensions.Telemetry/Http/HttpRouteParser.cs

```diff
@@ -38,61 +57,70 @@ public bool TryExtractParameters(
-                    // omitted parameter.
-                    string parameterValue = segment.DefaultValue;
+            int startIndex = segment.Start + offset;
 
-                    bool isRedacted = false;
+            // If we exceed a length of the http path it means that the appropriate http route
+            // has optional parameters or parameters with default values, and these parameters
+            // are omitted in the http path. In this case we return a default value of the
+            // omitted parameter.
+            string parameterValue = segment.DefaultValue;
 
-                    if (startIndex < httpPathAsSpan.Length)
-                    {
-                        var parameterContent = segment.Content;
-                        var parameterTemplateLength = parameterContent.Length + 2;
+            bool isRedacted = false;
 
-                        var length = httpPathAsSpan.Slice(startIndex).IndexOf(ForwardSlash);
+            if (startIndex < httpPathAsSpan.Length)
+            {
+                string parameterContent = segment.Content;
+                int parameterTemplateLength = parameterContent.Length + 2;
 
-                        if (segment.IsCatchAll || length == -1)
-                        {
-                            length = httpPathAsSpan.Slice(startIndex).Length;
-                        }
+                int length = httpPathAsSpan.Slice(startIndex).IndexOf(ForwardSlash);
 
-                        offset += length - parameterTemplateLength;
+                if (segment.IsCatchAll || length == -1)
+                {
+                    length = httpPathAsSpan.Slice(startIndex).Length;
+                }
 
-                        parameterValue = GetRedactedParameterValue(httpPathAsSpan, segment, startIndex, length, redactionMode, parametersToRedact, ref isRedacted);
-                    }
+                offset += length - parameterTemplateLength;
 
-                    httpRouteParameters[index++] = new HttpRouteParameter(segment.ParamName, parameterValue, isRedacted);
-                }
+                parameterValue = GetRedactedParameterValue(httpPathAsSpan, segment, startIndex, length, redactionMode, parametersToRedact, ref isRedacted);
             }
+
+            httpRouteParameters[index++] = new HttpRouteParameter(segment.ParamName, parameterValue, isRedacted);
         }
 
         return true;
     }
 
-    public ParsedRouteSegments ParseRoute(string httpRoute)
+    /// <summary>
+    /// Parses http route and breaks it into text and parameter segments.
+    /// </summary>
+    /// <param name="httpRoute">Http request's route template.</param>
+    /// <returns>Returns text and parameter segments of route.</returns>
+    public virtual ParsedRouteSegments ParseRoute(string httpRoute)
     {
-        return _routeTemplateSegmentsCache.GetOrAdd(httpRoute, httpRoute =>
+        return _routeTemplateSegmentsCache.GetOrAdd(httpRoute, _ =>
```

Comment:
```suggestion
        return _routeTemplateSegmentsCache.GetOrAdd(httpRoute, httpRoute =>
```

If we pass `httpRoute` as a parameter rather than getting it from the closure we'll save some bytes, won't we?